qiskit_experiments/visualization/plotters/base_plotter.py

```diff
@@ -0,0 +1,453 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Base plotter abstract class"""
+
+import warnings
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from qiskit_experiments.framework import Options
+from qiskit_experiments.visualization.drawers import BaseDrawer
+from qiskit_experiments.visualization.style import PlotStyle
+
+
+class BasePlotter(ABC):
+    """An abstract class for the serializable figure plotters.
+
+    A plotter takes data from an experiment analysis class or experiment and plots a given figure using a
+    drawing backend. Sub-classes define the kind of figure created and the expected data.
+
+    Data is split into series and figure data. Series data is grouped by series name (str). For
+    :class:`CurveAnalysis`, this is the model name for a curve fit. For series data associated with a
+    single series name and figure data, data-values are identified by a data-key (str). Different data
+    per series and figure must have a different data-key to avoid overwriting values. Experiment and
+    analysis results can be passed to the plotter so appropriate graphics can be drawn on the figure
+    canvas. Series data is added to the plotter using :meth:`set_series_data` whereas figure data is
+    added using :meth:`set_figure_data`. Series and figure data are retrieved using :meth:`data_for` and
+    :attr:`figure_data` respectively.
+
+    Series data contains values to be plotted on a canvas, such that the data can be grouped into subsets
+    identified by their series name. Series names can be thought of as legend labels for the plotted
+    data, and as curve names for a curve-fit. Figure data is not associated with a series or curve and is
+    instead only associated with the figure. Examples include analysis reports or other text that is
+    drawn onto the figure canvas.
+
+    # TODO: Add example usage and description of options and plot-options.
+    """
+
+    def __init__(self, drawer: BaseDrawer):
+        """Create a new plotter instance.
+
+        Args:
+            drawer: The drawer to use when creating the figure.
+        """
+        # Data to be plotted, such as scatter points, interpolated fits, and confidence intervals
+        self._series_data: Dict[str, Dict[str, Any]] = {}
```

Comment:
Series-data would also be used for QV, even if there is only one series name or the data-value is a single float (like a threshold value). Maybe there's a better name for `figure_data` to avoid that confusion?

I'm opposed to having abstract drawing shapes as the data-storage for `BasePlotter`, as this would further mix plotting code with analysis classes. Plotter subclasses are meant to be a bridge between analyses and the figure canvas, separating their responsibilities. However, I can definitely see a scenario a data-frame is used instead, as is mentioned in #893. A data-frame would have labels (data-keys) and indices (series-names) for values, which is all `BasePlotter` needs.

Reply:
I like the idea of renaming `figure_data` to `supplementary_data`. It makes sense as the core data being plotted is contained in `series_data`.

Reply:
Addressed in f1d1fc2.